core/src/main/scala/kafka/controller/PartitionStateMachineV2.scala

```diff
@@ -0,0 +1,356 @@
+      }
+    }
+    val (invalidPartitionsForElection, validPartitionsForElection) = leaderIsrAndControllerEpochPerPartition.partition { case (partition, leaderIsrAndControllerEpoch) =>
+      leaderIsrAndControllerEpoch.controllerEpoch > controllerContext.epoch
+    }
+    invalidPartitionsForElection.foreach { case (partition, leaderIsrAndControllerEpoch) =>
+      val failMsg = s"aborted leader election for partition $partition since the LeaderAndIsr path was " +
+        s"already written by another controller. This probably means that the current controller $controllerId went through " +
+        s"a soft failure and another controller was elected with epoch ${leaderIsrAndControllerEpoch.controllerEpoch}."
+      failedElections.put(partition, new StateChangeFailedException(failMsg))
+    }
+    if (validPartitionsForElection.isEmpty) {
+      return (Seq.empty, Seq.empty, failedElections.toMap)
+    }
+    val shuttingDownBrokers  = controllerContext.shuttingDownBrokerIds.toSet
+    val (partitionsWithoutLeaders, partitionsWithLeaders) = partitionLeaderElectionStrategy match {
+      case OfflinePartitionLeaderElectionStrategy =>
+        leaderForOffline(validPartitionsForElection).partition(_._2.isEmpty)
+      case ReassignPartitionLeaderElectionStrategy =>
+        leaderForReassign(validPartitionsForElection).partition(_._2.isEmpty)
+      case PreferredReplicaPartitionLeaderElectionStrategy =>
+        leaderForPreferredReplica(validPartitionsForElection).partition(_._2.isEmpty)
+      case ControlledShutdownPartitionLeaderElectionStrategy =>
+        leaderForControlledShutdown(validPartitionsForElection, shuttingDownBrokers).partition(_._2.isEmpty)
+    }
+    partitionsWithoutLeaders.foreach { case (partition, leaderAndIsrOpt, recipients) =>
+      val failMsg = s"Failed to elect leader for partition $partition under strategy $partitionLeaderElectionStrategy"
+      failedElections.put(partition, new StateChangeFailedException(failMsg))
+    }
+    val recipientsPerPartition = partitionsWithLeaders.map { case (partition, leaderAndIsrOpt, recipients) => partition -> recipients }.toMap
+    val adjustedLeaderAndIsrs = partitionsWithLeaders.map { case (partition, leaderAndIsrOpt, recipients) => partition -> leaderAndIsrOpt.get }.toMap
+    val (successfulUpdates, updatesToRetry, failedUpdates) = zkUtils.updateLeaderAndIsr(adjustedLeaderAndIsrs, controllerContext.epoch)
+    successfulUpdates.foreach { case (partition, leaderAndIsr) =>
+      val replicas = controllerContext.partitionReplicaAssignment(partition)
+      val leaderIsrAndControllerEpoch = LeaderIsrAndControllerEpoch(leaderAndIsr, controllerContext.epoch)
+      controllerContext.partitionLeadershipInfo.put(partition, leaderIsrAndControllerEpoch)
+      controllerBrokerRequestBatch.addLeaderAndIsrRequestForBrokers(recipientsPerPartition(partition), partition.topic, partition.partition, leaderIsrAndControllerEpoch, replicas)
```

Comment:
Structure-wise, it's probably better to do line 234 to 238 in electLeaderForPartitions()?

Reply:
We could, but in doing so, we'd have to return more state to `electLeaderForPartitions`.

Currently, `doElectLeaderForPartitions` returns `(Seq[TopicAndPartition], Seq[TopicAndPartition], Map[TopicAndPartition, Exception])` where the successes are just that first `Seq[TopicAndPartition]`.

If we delegate the work to `electLeaderForPartitions`, we'd have to pass back not only the successful partitions, but their LeaderAndIsr as well as the intended recipients of that partition's share of LeaderAndIsrRequest.

The code could end up messier as a result.

Reply:
Ok, we can leave it as it is then.